llvm/lib/Target/AMDGPU/AMDGPUTargetTransformInfo.cpp

```diff
@@ -75,6 +75,13 @@ static cl::opt<size_t> InlineMaxBB(
     cl::desc("Maximum number of BBs allowed in a function after inlining"
              " (compile time constraint)"));
 
+// This default unroll factor is based on microbenchmarks on gfx1030.
+static cl::opt<unsigned> MemcpyLoopUnroll(
+    "amdgpu-memcpy-loop-unroll",
+    cl::desc("Unroll factor (affecting 4x32-bit operations) to use for memory "
+             "operations when lowering memcpy as a loop, must be a power of 2"),
```

Comment:
Don't see why it needs to be a power of 2 (or where that's enforced)

Reply:
When it is not a power of two, the `(DL.getTypeStoreSize(LoopOpType) == DL.getTypeAllocSize(LoopOpType))` assertions that this patch introduces in LowerMemIntrinsics.cpp fail. The allocation sizes of these non-standard types are rounded up to the next power of two in our backend (because of their default alignment).
Mismatching alloc and store sizes are a problem because
- the gep-based offset computation in the memcpy/move lowering refers to the alloc size whereas
- how many bytes are read/written by the load/store operations is determined by the store size.

Is there a better way to enforce that constraint? I thought about changing the option to specify the exponent instead to avoid representing invalid state, but that seems not very intuitive from a usage perspective.

Reply:
Then the constraint is wrong, and the assertion is wrong. Non-power of 2 types are not a problem. <3 x i32> is important and there's nothing non-standard about it.

> the gep-based offset computation in the memcpy/move lowering refers to the alloc size whereas

I don't see where this is happening

Reply:
Then, something else must be broken, because using 12xi32 as a loop lowering type (with the assertions removed) gives me the following code on gfx1030 in the loop body:
```
.LBB0_1:                                ; %load-store-loop
                                        ; =>This Inner Loop Header: Depth=1
        s_clause 0x2
        flat_load_dwordx4 v[8:11], v[4:5]
        flat_load_dwordx4 v[12:15], v[4:5] offset:16
        flat_load_dwordx4 v[16:19], v[4:5] offset:32
        s_add_u32 s6, s6, 1
        s_addc_u32 s7, s7, 0
        v_add_co_u32 v4, vcc_lo, v4, 64
        v_cmp_lt_u64_e64 s4, s[6:7], 2
        v_add_co_ci_u32_e32 v5, vcc_lo, 0, v5, vcc_lo
        s_waitcnt vmcnt(2) lgkmcnt(2)
        flat_store_dwordx4 v[6:7], v[8:11]
        s_waitcnt vmcnt(1) lgkmcnt(2)
        flat_store_dwordx4 v[6:7], v[12:15] offset:16
        s_waitcnt vmcnt(0) lgkmcnt(2)
        flat_store_dwordx4 v[6:7], v[16:19] offset:32
        s_and_b32 vcc_lo, exec_lo, s4
        v_add_co_u32 v6, s4, v6, 64
        v_add_co_ci_u32_e64 v7, s4, 0, v7, s4
        s_cbranch_vccnz .LBB0_1
```
Notably, an iteration of this loop copies 12*4 = 48 bytes, but increments the memory addresses by 64. I think the separate-const-offset-from-gep pass does that, I'll investigate more.

Reply:
Correction: it's not separate-const-offset-from-gep, it is loop-reduce that introduces
```%scevgep = getelementptr i8, ptr %lsr.iv, i64 16```
while replacing
```%0 = getelementptr inbounds <3 x i32>, ptr %src, i64 %loop-index```

Reply:
So the GEPs are used incorrectly. You can always do the indexing in byte units. You don't need to preserve the type this way

Reply:
I opened #112707 to change the GEPs to i8. I'll rebase and update this PR once the other is landed.